### ISSUE TITLE
include-entry-exports per package

### ISSUE THREAD

**Issue by TkDodo** [webpro/knip#337]
We're in a monorepo that has both internal packages and packages (libs) that we both consume internally and publish.

We're currently not using `include-entry-exports` because it would flag false-positives for the exported packages.

One idea would be to allow to set that option "per package" so that we can define some packages to be excluded from that check.

Let me know what you think or if you have a better idea to address this, thanks.

**Comment by TkDodo** [comment]
I think we could re-structure from having everything in `packages` to something like `app` and `libs` and running knip twice, once against app with `--include-entry-points` and once against `libs` without 🤔 ...

**Comment by TkDodo** [comment]
what works is adding the entry files of libs we publish to `ignore`. This is good enough for me at the moment because it's only 2 libs. So feel free to close this if you don't like the suggestion because it has a workaround :)

**Comment by webpro** [comment]
Thanks Dominik, I like the suggestion (to configure it on workspace level). I think it's a fairly common use case. I'm probably going to include it soon in the upcoming v3.

**Comment by TkDodo** [comment]
> what works is adding the entry files of libs we publish to ignore. 

hmm I'm circling back on that, because that doesn't work. Maybe I'm misunderstanding what `ignore` is doing? I guess this also comes back to this question:

- #340 

basically, the generated export reads:

```
Unused exports (2)
someExport           packages/my-lib/src/index.ts
someOtherExport      packages/my-lib/src/index.ts
```

And I'd like to say: Oh yeah, this is one of the libs we export, so please ignore all errors from `packages/my-lib/src/index.ts` because that's actually a public entry point.

but when I add:

```
ignore: [
  'packages/my-lib/src/index.ts',
]
```

the errors still show up 🤔 . Not sure where I'm going wrong here. I guess I could write a [preprocessor](https://github.com/webpro/knip/blob/main/docs/reporters-and-preprocessors.md#preprocessors) for the report and exclude those lines, but not sure if I'm going somewhere horribly wrong here 😅

**Comment by webpro** [comment]
A few things about this:

- The `ignore` is a filter applied to what goes into the program, it does not filter results
- I totally understand this is very confusing, also when looking at the other options that came later:
  - `ignoreBinaries` and `ignoreDependencies` are actually filters on the results
  - `ignoreWorkspaces` is a filter on the input
- The `ignore` option should work better when placed inside the workspace config (where this becomes `ignore: ['src/index.ts']` in the `packages/my-lib` workspace config.
- You can use `--debug` to see the glob actions taken, including the ignored items and the resulting files that go into the program.
- Eventually there's lists of resolved files in the debug output, and here's the thing: the `ignore` config has no influence at all on the module resolution and the files TS adds to the program. It prevents entry files from initially being added, but if TS decides it's imported somewhere and makes it part of the program then currently it has the last word on this.

So, the least I can do is document this better, but it also needs a bit more thinking through :)

**Comment by webpro** [comment]
So actually yeah, at this point, if moving the `ignore` option to the workspace config doesn't help, that preprocessor is a way out if you really need it.

**Comment by webpro** [comment]
:rocket: _This issue has been resolved in v3.0.0-canary.2. See [Release 3.0.0-canary.2](https://github.com/webpro/knip/releases/tag/3.0.0-canary.2) for release notes._
